File: bin/tests/system/fetchlimit/ans4/ans.pl
#!/usr/bin/perl -w
#
# Copyright (C) 2015, 2016, 2018  Internet Systems Consortium, Inc. ("ISC")
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

#
# Don't respond if the "norespond" file exists; otherwise respond to
# any A or AAAA query.
#

use IO::File;
use IO::Socket;
use Net::DNS;
use Net::DNS::Packet;

my $sock = IO::Socket::INET->new(LocalAddr => "10.53.0.4",
   LocalPort => 5300, Proto => "udp") or die "$!";

my $pidf = new IO::File "ans.pid", "w" or die "cannot open pid file: $!";
print $pidf "$$\n" or die "cannot write pid file: $!";
$pidf->close or die "cannot close pid file: $!";
sub rmpid { unlink "ans.pid"; exit 1; };

$SIG{INT} = \&rmpid;
$SIG{TERM} = \&rmpid;

for (;;) {
	$sock->recv($buf, 512);

	print "**** request from " , $sock->peerhost, " port ", $sock->peerport, "\n";

	my $packet;

	if ($Net::DNS::VERSION > 0.68) {
		$packet = new Net::DNS::Packet(\$buf, 0);
		$@ and die $@;
	} else {
		my $err;
		($packet, $err) = new Net::DNS::Packet(\$buf, 0);
		$err and die $err;
	}

	print "REQUEST:\n";
	$packet->print;

	$packet->header->qr(1);

	my @questions = $packet->question;
	my $qname = $questions[0]->qname;
	my $qtype = $questions[0]->qtype;

	my $donotrespond = 0;

	if (-e 'norespond') {
		$donotrespond = 1;
	} else {
		$packet->header->aa(1);
		if ($qtype eq "A") {
			$packet->push("answer",
				      new Net::DNS::RR($qname .
						       " 300 A 192.0.2.1"));
		} elsif ($qtype eq "AAAA") {
			$packet->push("answer",
				      new Net::DNS::RR($qname .
						" 300 AAAA 2001:db8:beef::1"));
		}
	}

	if ($donotrespond == 0) {
		$sock->send($packet->data);
		print "RESPONSE:\n";
		$packet->print;
		print "\n";
	}
}

File: bin/tests/system/notify/ans7/ans.py
# Copyright (C) Internet Systems Consortium, Inc. ("ISC")
#
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0.  If a copy of the MPL was not distributed with this
# file, you can obtain one at https://mozilla.org/MPL/2.0/.
#
# See the COPYRIGHT file distributed with this work for additional
# information regarding copyright ownership.

import os
import select
import signal
import socket
import sys
import time

import dns.flags
import dns.message


def port():
    env_port = os.getenv("PORT")
    if env_port is None:
        env_port = 5300
    else:
        env_port = int(env_port)

    return env_port


def udp_listen(port):
    udp = socket.socket(type=socket.SOCK_DGRAM)
    udp.bind(("10.53.0.7", port))

    return udp


def tcp_listen(port):
    tcp = socket.socket(type=socket.SOCK_STREAM)
    tcp.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    tcp.bind(("10.53.0.7", port))
    tcp.listen(100)

    return tcp


def udp_response(udp):
    qrybytes, clientaddr = udp.recvfrom(65535)
    qry = dns.message.from_wire(qrybytes)
    answbytes = b'utter garbage'
    udp.sendto(answbytes, clientaddr)


def tcp_response(tcp):
    csock, _clientaddr = tcp.accept()
    while True:
      qrylen = csock.recv(4096)
      answbytes = b'utter garbage\n'
      csock.send(answbytes)
    csock.close()


def sigterm(signum, frame):
    os.remove("ans.pid")
    sys.exit(0)


def write_pid():
    with open("ans.pid", "w") as f:
        pid = os.getpid()
        f.write("{}".format(pid))


signal.signal(signal.SIGTERM, sigterm)
write_pid()

udp = udp_listen(port())
tcp = tcp_listen(port())

input = [udp, tcp]

while True:
    try:
        inputready, outputready, exceptready = select.select(input, [], [])
    except select.error:
        break
    except socket.error:
        break
    except KeyboardInterrupt:
        break

    for s in inputready:
        if s == udp:
            udp_response(udp)
        if s == tcp:
            tcp_response(tcp)

sigterm(signal.SIGTERM, 0)
